static/app/utils/profiling/flamegraph/selectNearestFrame.ts

```diff
@@ -0,0 +1,94 @@
+import {FlamegraphFrame} from '../flamegraphFrame';
+
+export type DirectionX = 'left' | 'right';
+export type DirectionY = 'up' | 'down';
+export type Direction = DirectionY | DirectionX;
+/**
+ * selectNearestFrame walks a FlamegraphFrame tree the direction specified and returns the nearest
+ * FlamegraphFrame relative to the starting node.
+ *
+ * below is a gist of the walking algorithm:
+ *
+ *  up    => frame's parent
+ *  down  => frame's first child
+ *  left  =>
+ *          |> walk up to find a parent that contains the current frame w/ a sibling to the left
+ *          |> walk down the sibling tree following right branches
+ *          |> return node matching target depth, if max depth found is less than target return
+ *  right => same as left, but we find the right sibling, and follow left branches down until target/max depth is reached
+ */
+export function selectNearestFrame(n: FlamegraphFrame, dir: Direction) {
+  if (!n) {
+    return null;
+  }
+
+  const targetDepth = n.depth;
+  let parent = n.parent;
+  let node = n;
+
+  if (dir === 'up' && parent) {
+    return parent;
+  }
+
+  const child = n.children?.[0];
+  if (dir === 'down' && child) {
+    return child;
+  }
```

Comment:
When the direction is down and there is no child this falls through and behaves like a left. Is this intended?

Reply:
nope, that's a bug!

Reply:
Can you also add some tests for this too? 😄

Reply:
done!